File: macos/Sources/FieldCommand/Terrain.swift
import AppKit

/// Organic terrain rendering, a port of the Linux edition's terrain.py: water and cliffs are drawn from a blurred,
/// noise-warped mask of the map's wall rectangles, so overlapping pieces merge into natural lakes, rivers and
/// ridges. Collision still uses the rects; the drawn shape covers them and extends a little beyond.
enum Terrain {
    static let res = 2.0  // world units per pixel

    struct Wall { let rect: CGRect; let water: Bool }

    static func walls(_ map: [String: Any]) -> [Wall] {
        jArr(map["walls"]).map { w in
            let v = jArr(w)
            return Wall(rect: CGRect(x: jNum(v[0]), y: jNum(v[1]), width: jNum(v[2]) - jNum(v[0]), height: jNum(v[3]) - jNum(v[1])),
                        water: jStr(v[4]) == "water")
        }
    }

    static func bridges(_ map: [String: Any]) -> [CGRect] {
        jArr(map["bridges"]).map { b in
            let v = jArr(b)
            return CGRect(x: jNum(v[0]), y: jNum(v[1]), width: jNum(v[2]) - jNum(v[0]), height: jNum(v[3]) - jNum(v[1]))
        }
    }

    // MARK: Field helpers (row 0 = top of the map)

    private struct Field {
        let w: Int, h: Int
        var a: [Float]
        init(_ w: Int, _ h: Int) { self.w = w; self.h = h; a = [Float](repeating: 0, count: w * h) }
    }

    private static func mask(_ rects: [CGRect], _ w: Int, _ h: Int) -> Field {
        var f = Field(w, h)
        let H = Double(worldSize.height)
        for r in rects {
            let c0 = max(0, Int(Double(r.minX) / res)), c1 = min(w, Int((Double(r.maxX) / res).rounded(.up)))
            let r0 = max(0, Int((H - Double(r.maxY)) / res)), r1 = min(h, Int(((H - Double(r.minY)) / res).rounded(.up)))
            guard c0 < c1, r0 < r1 else { continue }
            for y in r0..<r1 { for x in c0..<c1 { f.a[y * w + x] = 1 } }
        }
        return f
    }

    /// Separable box blur, repeated to approximate a gaussian.
    private static func blur(_ f: inout Field, _ r: Int, passes: Int = 3) {
        let w = f.w, h = f.h
        var tmp = [Float](repeating: 0, count: max(w, h) + 2 * r + 2)
        let norm = 1 / Float(2 * r + 1)
        for _ in 0..<passes {
            for y in 0..<h {  // horizontal
                var s: Float = 0
                let row = y * w
                for i in -r - 1..<r { s += f.a[row + min(w - 1, max(0, i))] }
                for x in 0..<w {
                    s += f.a[row + min(w - 1, x + r)] - f.a[row + max(0, x - r - 1)]
                    tmp[x] = s * norm
                }
                for x in 0..<w { f.a[row + x] = tmp[x] }
            }
            for x in 0..<w {  // vertical
                var s: Float = 0
                for i in -r - 1..<r { s += f.a[min(h - 1, max(0, i)) * w + x] }
                for y in 0..<h {
                    s += f.a[min(h - 1, y + r) * w + x] - f.a[max(0, y - r - 1) * w + x]
                    tmp[y] = s * norm
                }
                for y in 0..<h { f.a[y * w + x] = tmp[y] }
            }
        }
    }

    /// Smooth value noise in [-1, 1]; `period` is the feature size in pixels.
    private static func noise(_ w: Int, _ h: Int, period: Double, octaves: Int, seed: UInt64) -> Field {
        var f = Field(w, h)
        var rng = SeededRNG(seed)
        var amp: Float = 0.5, norm: Float = 0, p = period
        for _ in 0..<octaves {
            let gw = Int(Double(w) / p) + 2, gh = Int(Double(h) / p) + 2
            let lattice = (0..<(gw * gh)).map { _ in Float.random(in: 0...1, using: &rng) }
            for y in 0..<h {
                let fy = Double(y) / p, y0 = Int(fy)
                var ty = Float(fy - Double(y0)); ty = ty * ty * (3 - 2 * ty)
                for x in 0..<w {
                    let fx = Double(x) / p, x0 = Int(fx)
                    var tx = Float(fx - Double(x0)); tx = tx * tx * (3 - 2 * tx)
                    let a = lattice[y0 * gw + x0], b = lattice[y0 * gw + x0 + 1]
                    let c = lattice[(y0 + 1) * gw + x0], d = lattice[(y0 + 1) * gw + x0 + 1]
                    let top = a + (b - a) * tx, bottom = c + (d - c) * tx
                    f.a[y * w + x] += (top + (bottom - top) * ty) * amp
                }
            }
            norm += amp
            amp *= 0.5
            p /= 2
        }
        for i in f.a.indices { f.a[i] = f.a[i] / norm * 2 - 1 }
        return f
    }

    @inline(__always) private static func smooth(_ e0: Float, _ e1: Float, _ x: Float) -> Float {
        let t = min(1, max(0, (x - e0) / (e1 - e0)))
        return t * t * (3 - 2 * t)
    }

    // MARK: Rendering

    /// A world-sized (at 1/`res` scale) image of all water, cliffs and bridges, or nil when the map has none.
    static func image(_ map: [String: Any]) -> CGImage? {
        let walls = walls(map), bridges = bridges(map)
        guard !walls.isEmpty else { return nil }
        let w = Int(Double(worldSize.width) / res), h = Int(Double(worldSize.height) / res)
        let seed = UInt64(max(1, jInt(map["seed"])))
        let warp = noise(w, h, period: 150 / res, octaves: 3, seed: seed &+ 11)
        let fine = noise(w, h, period: 40 / res, octaves: 2, seed: seed &+ 12)
        let patchy = noise(w, h, period: 30, octaves: 2, seed: seed &+ 13)
        // Premultiplied colour and coverage, 0...1
        var rgb = [Float](repeating: 0, count: w * h * 3)
        var alpha = [Float](repeating: 0, count: w * h)

        let water = walls.filter { $0.water }.map { $0.rect }
        if !water.isEmpty {
            var f = mask(water + bridges, w, h)
            blur(&f, 16)
            for y in 0..<h {
                let ripplePhase = Float(y) * 0.5
                for x in 0..<w {
                    let i = y * w + x
                    let b = f.a[i]
                    guard b > 0.001 else { continue }
                    let v = b + warp.a[i] * 0.24 * smooth(0, 0.2, b)
                    let shore = smooth(0.10, 0.17, v)
                    guard shore > 0 else { continue }
                    let wet = smooth(0.28, 0.33, v)
                    let depth = smooth(0.35, 0.95, v + fine.a[i] * 0.04)
                    // sand fading to damp mud at the waterline
                    let mud = smooth(0.18, 0.28, v)
                    var cr = 118 + fine.a[i] * 14, cg = 104 + fine.a[i] * 14, cb = 70 + fine.a[i] * 14
                    cr += (70 - cr) * mud; cg += (58 - cg) * mud; cb += (36 - cb) * mud
                    var wr = 46 + (16 - 46) * depth, wg = 112 + (50 - 112) * depth, wb = 128 + (76 - 128) * depth
                    let ripple = sin(ripplePhase + Float(x) * 0.04 + fine.a[i] * 2.5) * 0.5 + 0.5
                    let rp = pow(ripple, 14) * smooth(0, 0.6, patchy.a[i]) * 30 * (0.4 + depth)
                    let v2 = (v - 0.315) / 0.018
                    let foam = exp(-v2 * v2) * 60
                    wr += rp + foam; wg += rp + foam; wb += rp + foam
                    cr += (wr - cr) * wet; cg += (wg - cg) * wet; cb += (wb - cb) * wet
                    rgb[i * 3] = cr * shore; rgb[i * 3 + 1] = cg * shore; rgb[i * 3 + 2] = cb * shore
                    alpha[i] = shore * 0.97
                }
            }
        }

        let cliffs = walls.filter { !$0.water }.map { $0.rect }
        if !cliffs.isEmpty {
            var f = mask(cliffs, w, h)
            blur(&f, 10)
            var v = f.a, hgt = f.a
            for i in v.indices {
                let b = f.a[i]
                v[i] = b + warp.a[i] * 0.22 * smooth(0, 0.2, b)
                hgt[i] = smooth(0.22, 0.75, v[i]) * 40 + (fine.a[i] * 9 + abs(warp.a[i]) * 8) * smooth(0.2, 0.5, v[i])
            }
            var body = [Float](repeating: 0, count: w * h)
            for i in v.indices { body[i] = smooth(0.24, 0.30, v[i]) }
            for y in 0..<h {
                for x in 0..<w {
                    let i = y * w + x
                    // cast shadow down-right
                    let sy = y - 7, sx = x - 5
                    let caster: Float = sy >= 0 && sx >= 0 ? body[sy * w + sx] : 0
                    let sh = caster * (1 - body[i]) * 0.4
                    if sh > 0 {
                        rgb[i * 3] *= 1 - sh; rgb[i * 3 + 1] *= 1 - sh; rgb[i * 3 + 2] *= 1 - sh
                        alpha[i] = max(alpha[i], sh)
                    }
                    let bd = body[i]
                    guard bd > 0 else { continue }
                    // height field -> lighting (light from the top-left of the screen)
                    let gx = (hgt[y * w + min(w - 1, x + 1)] - hgt[y * w + max(0, x - 1)]) * 0.5
                    let gy = (hgt[min(h - 1, y + 1) * w + x] - hgt[max(0, y - 1) * w + x]) * 0.5
                    let light = min(1.25, max(0.15, 0.62 + gx * 0.20 + gy * 0.26))
                    let top = smooth(0.55, 0.75, v[i])
                    let n = fine.a[i] * 26
                    let r = (92 + 36 * top + n) * light, g = (86 + 36 * top + n) * light, b = (76 + 32 * top + n) * light
                    rgb[i * 3] = rgb[i * 3] * (1 - bd) + r * bd
                    rgb[i * 3 + 1] = rgb[i * 3 + 1] * (1 - bd) + g * bd
                    rgb[i * 3 + 2] = rgb[i * 3 + 2] * (1 - bd) + b * bd
                    alpha[i] = max(alpha[i], bd)
                }
            }
        }

        var px = [UInt8](repeating: 0, count: w * h * 4)
        for i in 0..<(w * h) {
            let a = min(1, alpha[i])
            // stored premultiplied: colour already weighted by coverage, capped to alpha
            px[i * 4] = UInt8(min(a * 255, max(0, rgb[i * 3])))
            px[i * 4 + 1] = UInt8(min(a * 255, max(0, rgb[i * 3 + 1])))
            px[i * 4 + 2] = UInt8(min(a * 255, max(0, rgb[i * 3 + 2])))
            px[i * 4 + 3] = UInt8(a * 255)
        }
        return px.withUnsafeMutableBytes { buf -> CGImage? in
            guard let ctx = CGContext(data: buf.baseAddress, width: w, height: h, bitsPerComponent: 8, bytesPerRow: w * 4,
                                      space: Art.srgb, bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue) else { return nil }
            // Bridges, drawn in world coordinates (the context's origin is bottom-left, y up).
            ctx.scaleBy(x: CGFloat(1 / res), y: CGFloat(1 / res))
            for (k, r) in bridges.enumerated() { drawBridge(ctx, r, seed: UInt64(k + 7)) }
            return ctx.makeImage()
        }
    }

    private static func drawBridge(_ ctx: CGContext, _ r: CGRect, seed: UInt64) {
        var rng = SeededRNG(seed)
        let deck = r.insetBy(dx: 0, dy: 6)
        Art.fill(ctx, Art.rr(deck.offsetBy(dx: 6, dy: -8), 4), .rgb(0, 0, 0, 0.35))
        Art.linear(ctx, Art.rr(deck, 4), [.rgb(0.58, 0.43, 0.26), .rgb(0.44, 0.31, 0.18)],
                   CGPoint(x: deck.midX, y: deck.maxY), CGPoint(x: deck.midX, y: deck.minY))
        ctx.setStrokeColor(NSColorRGB(0.25, 0.17, 0.09, 0.8))
        ctx.setLineWidth(1.4)
        var x = deck.minX + 12
        while x < deck.maxX {
            ctx.move(to: CGPoint(x: x, y: deck.minY + 1))
            ctx.addLine(to: CGPoint(x: x, y: deck.maxY - 1))
            x += 12 + CGFloat.random(in: -1...1, using: &rng)
        }
        ctx.strokePath()
        for y in [deck.minY + 2, deck.maxY - 2] {
            let rail = CGRect(x: deck.minX, y: y - 4, width: deck.width, height: 8)
            Art.fill(ctx, Art.rr(rail, 3), .rgb(0.40, 0.28, 0.15))
            Art.fill(ctx, Art.rr(rail.insetBy(dx: 1, dy: 3).offsetBy(dx: 0, dy: 1.5), 1.5), .rgb(0.55, 0.40, 0.24))
            var px = deck.minX + 8
            while px < deck.maxX {
                Art.fill(ctx, Art.circle(CGPoint(x: px, y: y), 3.2), .rgb(0.22, 0.15, 0.08))
                px += 40
            }
        }
        Art.stroke(ctx, Art.rr(deck, 4), .rgb(0.20, 0.13, 0.06, 0.9), 1.5)
    }

    private static func NSColorRGB(_ r: CGFloat, _ g: CGFloat, _ b: CGFloat, _ a: CGFloat) -> CGColor {
        CGColor(srgbRed: r, green: g, blue: b, alpha: a)
    }
}

File: linux/fieldcommand/terrain.py
"""Organic terrain rendering: water and cliffs are drawn from a blurred, noise-warped mask of the map's wall
rectangles, so overlapping pieces merge into natural lakes, rivers and ridges. Collision still uses the rects;
the drawn shape always covers them and extends a little beyond.
"""
import math

import numpy as np
import pygame

from .art import _value_noise
from . import defs

RES = 2  # world units per pixel of the working buffers (mega maps use a coarser buffer, see _res)


def _res():
    """Keep the working buffers about 2000 x 1400 whatever the map size, so big maps cost no more memory."""
    return max(RES, int(math.ceil(defs.WORLD_W / 2000)))


def _mask(rects, W, H, res):
    m = np.zeros((H, W), np.float32)
    for x0, y0, x1, y1 in rects:
        # rect in world coords (y up) -> buffer rows (y down)
        c0, c1 = int(max(0, x0 // res)), int(min(W, -(-x1 // res)))
        r0, r1 = int(max(0, (defs.WORLD_H - y1) // res)), int(min(H, -(-(defs.WORLD_H - y0) // res)))
        m[r0:r1, c0:c1] = 1.0
    return m


def _blur(a, r, passes=3):
    """Separable box blur, repeated to approximate a gaussian."""
    for _ in range(passes):
        for axis in (0, 1):
            p = np.pad(a, [(r + 1, r) if ax == axis else (0, 0) for ax in range(2)], mode="edge")
            c = np.cumsum(p, axis=axis, dtype=np.float32)
            if axis == 0:
                a = (c[2 * r + 1:] - c[:-2 * r - 1]) / (2 * r + 1)
            else:
                a = (c[:, 2 * r + 1:] - c[:, :-2 * r - 1]) / (2 * r + 1)
    return a


def _noise(W, H, period, octaves, seed):
    """Smooth value noise in [-1, 1]; `period` is the feature size in buffer pixels."""
    n = max(W, H)
    base = _value_noise(n, max(2, min(int(n / period), 256 >> (octaves - 1))), octaves, seed)
    return base[:H, :W] * 2 - 1


def _smooth(e0, e1, x):
    t = np.clip((x - e0) / (e1 - e0), 0, 1)
    return t * t * (3 - 2 * t)


def _lerp(a, b, t):
    return a + (np.asarray(b, np.float32) - a) * t[..., None]


def render(map_spec):
    """Returns a world-sized RGBA surface of all water, cliffs and high ground (or None when the map has none)."""
    walls = map_spec.get("walls", [])
    ridges = map_spec.get("ridges", [])
    if not walls and not ridges:
        return None
    res = _res()
    W, H = int(defs.WORLD_W) // res, int(defs.WORLD_H) // res
    seed = int(map_spec.get("seed", 1))
    rgb = np.zeros((H, W, 3), np.float32)
    alpha = np.zeros((H, W), np.float32)
    warp = _noise(W, H, 150 / res, 3, seed + 11)  # low-frequency wobble of the outlines
    fine = _noise(W, H, 40 / res, 2, seed + 12)

    water = [w[:4] for w in walls if w[4] == "water"] + [tuple(b) for b in map_spec.get("bridges", [])]
    if any(w[4] == "water" for w in walls):
        b = _blur(_mask(water, W, H, res), max(6, int(32 / res)))
        f = b + warp * 0.24 * _smooth(0.0, 0.2, b)  # wobble only near the water
        shore = _smooth(0.10, 0.17, f)
        wet = _smooth(0.28, 0.33, f)
        depth = _smooth(0.35, 0.95, f + fine * 0.04)
        sand = np.array([118, 104, 70], np.float32) + fine[..., None] * 14
        col = _lerp(sand, (70, 58, 36), _smooth(0.18, 0.28, f))  # damp mud next to the water
        shallow = np.array([46, 112, 128], np.float32)
        deep = np.array([16, 50, 76], np.float32)
        wc = _lerp(np.broadcast_to(shallow, col.shape).copy(), deep, depth)
        ripple = np.sin(np.arange(H)[:, None] * 0.5 + np.arange(W)[None, :] * 0.04 + fine * 2.5) * 0.5 + 0.5
        patchy = _smooth(0.0, 0.6, _noise(W, H, 30, 2, seed + 13))
        wc += (ripple ** 14 * patchy)[..., None] * 30 * (0.4 + depth[..., None])
        foam = np.exp(-((f - 0.315) / 0.018) ** 2)
        wc += foam[..., None] * 60
        col = col * (1 - wet[..., None]) + wc * wet[..., None]
        rgb = rgb * (1 - shore[..., None]) + col * shore[..., None]
        alpha = np.maximum(alpha, shore * 0.97)

    cliffs = [w[:4] for w in walls if w[4] == "cliff"]
    if cliffs:
        b = _blur(_mask(cliffs, W, H, res), max(5, int(20 / res)))
        f = b + warp * 0.22 * _smooth(0.0, 0.2, b)
        body = _smooth(0.24, 0.30, f)
        # height field -> lighting (light from the top-left of the screen); fine noise makes it craggy
        hgt = _smooth(0.22, 0.75, f) * 40 + (fine * 9 + np.abs(warp) * 8) * _smooth(0.2, 0.5, f)
        gy, gx = np.gradient(hgt)
        light = np.clip(0.62 + gx * 0.20 + gy * 0.26, 0.15, 1.25)
        top = _smooth(0.55, 0.75, f)
        rock = _lerp(np.broadcast_to(np.array([92, 86, 76], np.float32), f.shape + (3,)).copy(), (128, 122, 108), top)
        rock += fine[..., None] * 26
        rock = rock * light[..., None]
        # cast shadow down-right
        sh = np.roll(np.roll(body, 7, axis=0), 5, axis=1) * (1 - body)
        rgb = rgb * (1 - sh[..., None] * 0.4)
        alpha = np.maximum(alpha, sh * 0.4)
        rgb = rgb * (1 - body[..., None]) + rock * body[..., None]
        alpha = np.maximum(alpha, body)

    if ridges:
        # High ground: a raised plateau of drier grass with a lit rim toward the light and a soft cast shadow
        # down-right, so it reads as a step up rather than a patch of colour.
        b = _blur(_mask([tuple(r[:4]) for r in ridges], W, H, res), max(4, int(14 / res)))
        f = b + warp * 0.12 * _smooth(0.0, 0.3, b)
        body = _smooth(0.36, 0.44, f)
        hgt = _smooth(0.32, 0.62, f) * 18 + fine * 3 * _smooth(0.3, 0.6, f)
        gy, gx = np.gradient(hgt)
        light = np.clip(0.80 + gx * 0.30 + gy * 0.36, 0.45, 1.35)
        grass = np.broadcast_to(np.array([112, 126, 66], np.float32), f.shape + (3,)).copy() + fine[..., None] * 12
        grass = grass * light[..., None]
        sh = np.roll(np.roll(body, 4, axis=0), 3, axis=1) * (1 - body)
        rgb = rgb * (1 - sh[..., None] * 0.35)
        alpha = np.maximum(alpha, sh * 0.35)
        rgb = rgb * (1 - body[..., None]) + grass * body[..., None]
        alpha = np.maximum(alpha, body)

    # rgb currently holds colour already weighted by coverage for cliffs' shadow; un-premultiply
    out = np.zeros((H, W, 4), np.uint8)
    a = np.clip(alpha, 1e-4, 1)
    out[..., :3] = np.clip(rgb / a[..., None], 0, 255).astype(np.uint8)
    out[..., 3] = (np.clip(alpha, 0, 1) * 255).astype(np.uint8)
    surf = pygame.image.frombuffer(out.tobytes(), (W, H), "RGBA")
    return pygame.transform.smoothscale(surf, (int(defs.WORLD_W), int(defs.WORLD_H)))
